### PR TITLE
Update policy add-on to work with crd OwnerReferences

### PR DESCRIPTION
**Pull request by EmandM**

### CODE REVIEW

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/gatekeeper.yaml [hunk 1]
```diff
@@ -320,7 +320,7 @@ spec:
           valueFrom:
             fieldRef:
               fieldPath: metadata.name
-        image: mcr.microsoft.com/oss/open-policy-agent/gatekeeper:v3.1.0-beta.6
+        image: mcr.microsoft.com/oss/open-policy-agent/gatekeeper:v3.1.0-beta.7
```
Comment (ritazh) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/gatekeeper.yaml]: whenever we upgrade the image tag, please also make sure the yaml has all the changes as well. I see it's currently missing few things.

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+spec:
+  template:
+    spec:
+      serviceAccountName: pre-delete-admin
+      automountServiceAccountToken: true 
+      restartPolicy: Never
+      terminationGracePeriodSeconds: 0
+      containers:
+        - name: pre-delete
+          image: "appropriate/curl" # alpine + curl
```
Comment (ritazh) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: we probably need one built by microsoft

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          imagePullPolicy: IfNotPresent
+          command: 
+            - "/bin/sh"
+            - "-ec"
+            - |
+              set -o pipefail
+              TOKEN=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
+              curl -s -X DELETE -k https://kubernetes.default.svc/apis/apps/v1/namespaces/kube-system/deployments/azure-policy \
```
Comment (ritazh) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: the namespace should probably be a variable

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-serviceaccount.yaml [hunk 4]
```diff
@@ -15,6 +15,9 @@ rules:
 - apiGroups: ["templates.gatekeeper.sh"]
   resources: ["constrainttemplates"]
   verbs: ["create", "delete", "update", "list", "get"]
+- apiGroups: [""]
+  resources: ["pods"]
+  verbs: ["get"]
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-serviceaccount.yaml]: Does this mean Azure-policy has permissions to get info on all pods? Would customer be fine with giving that level of information? Can we just restrict to a namespace or a particular pod?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-deploy.yaml [hunk 5]
```diff
@@ -46,17 +46,19 @@ spec:
...
-        - name: AUDIT_INTERVAL
-          value: 5m   # Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h"
-        - name: TELEMETRY_HEARTBEAT_INTERVAL
-          value: 5m   # Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h"
         - name: CLIENT_SECRET
           valueFrom:
             secretKeyRef:
               name: azure-policy
               key: client-secret
+        - name: POD_NAME
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-deploy.yaml:59]: Need to remove INTERVALs in Azure Policy helm chart that Ruchika has currently for AKS. Can you please follow up with Ruchika on this?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: pre-delete-agent
+  annotations:
+    "helm.sh/hook": pre-install
+    "helm.sh/hook-weight": "20"
+rules:
+- apiGroups: ["apps"]
+  resources: ["deployments"]
+  verbs: ["delete"]
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: Does this means azure-policy account can remove any deployment? Would this be fine for customer? Can we confirm on this? Can we restrict the permissions to particular pod/deployment and a namespace?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: pre-delete-admin
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: Can this be named something like azure-policy-delete-admin?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  namespace: kube-system
+  annotations:
+    "helm.sh/hook": pre-install
+    "helm.sh/hook-weight": "20"
+---
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: pre-delete-agent
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: Can this be named something like azure-policy-pre-delete-agent?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    "helm.sh/hook-delete-policy": "hook-succeeded"
+spec:
+  template:
+    spec:
+      serviceAccountName: pre-delete-admin
+      automountServiceAccountToken: true 
+      restartPolicy: Never
+      terminationGracePeriodSeconds: 0
+      containers:
+        - name: pre-delete
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: Can you rename to azure-policy-pre-delete?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml (hunk 3 above)
Comment (RamyasreeChakka) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: @EmandM Can you confirm with AKS team if calling Kubernetes service using https://kubernetes.default.svc is a supported way. I think it should be supported. Just to be double sure that the dns is mapped to api server on managed Kubernetes as well.

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+- kind: ServiceAccount
+  name: pre-delete-admin
+  namespace: kube-system
+roleRef:
+  kind: Role
+  name: pre-delete-agent
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: batch/v1
+kind: Job
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: I see this pre-delete hook job is removing the azure-policy pod which has been deployed through helm chart? Would there be any race conditions in below scenarios...
1. After pre-delete hook execution, helm tries to uninstall the pod but the pod has already been deleted. Would this cause any error?
2. Azure-Policy pod got deleted but helm hasn't yet finished uninstall. Before chart is uninstalled, would there be a scenario where Kubernetes master server sees pod missing and tries to create a new pod?

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml (hunk 3 above)
Comment (EmandM) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: https://kubernetes.default.svc is the default way to access the kubernetes api server from inside any container running on a kubernetes cluster. That doesn't change even if the api server is managed. The permissions for this api are managed via service accounts, which is why we install a service account along with the hook.

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-deploy.yaml (hunk 5 above)
Comment (EmandM) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-deploy.yaml:59]: Is there any rush to remove the intervals from AKS? The reconcile loop ensures the intervals can't be changed by the user. If there's no rush, can we put the change in with Gatekeeper v3 changes?

Comment (RamyasreeChakka) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-deploy.yaml:59]: Fine with me.

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml (hunk 6 above)
Comment (EmandM) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: Only the hook has permissions to delete deployments. It uses a separate service account to ensure the add-on does not have these permissions

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-serviceaccount.yaml (hunk 4 above)
Comment (EmandM) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/azure-policy-serviceaccount.yaml]: I've updated so the azure-policy only has permissions to get info on pods in the `kube-system` namespace. Unfortunately we can't restrict it any tighter than that.

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml (hunk 3 above)
Comment (EmandM) [reply Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: Updated namespace to be a variable

File: Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  namespace: kube-system
+roleRef:
+  kind: Role
+  name: azure-policy-delete-agent
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: addon-delete
```
Comment (RamyasreeChakka) [inline Kubernetes/helmcharts/azure-policy-addon-arc-connected-clusters/templates/remove-addon.yaml]: nit: azure-policy-addon-delete

**Review by RamyasreeChakka — APPROVED** [review]

**Pull request merged**